src/genotyper.cc

```diff
@@ -41,6 +41,286 @@ static bool is_pseudo_ref_record(const bcf_hdr_t* hdr, bcf1_t* record) {
     return true;
 }
 
+class IFormatFieldHelper {
+
+public:
+    // basic information for the retained field
+    retained_format_field field_info;
+
+
+    // Expected number of samples in output bcf record
+    int n_samples;
+
+    // Expected number of values per sample in the **output** (ie unified site)
+    // Note this may differ from the count of input for RetainedFieldNumber::ALLELES
+    // and RetainedFieldNumber::ALT since the number of alleles may differ in input and
+    // output
+    int count;
+
+     IFormatFieldHelper(const retained_format_field field_info_, int n_samples_, int count_) : field_info(field_info_), n_samples(n_samples_), count(count_) {}
+
+     IFormatFieldHelper() = default;
+
+     virtual Status add_record_data(const string& dataset, const bcf_hdr_t* dataset_header, bcf1_t* record,
+                                    const map<int, int>& sample_mapping, const vector<int> allele_mapping,
+                                    const vector<string>& field_names={}, int n_val_per_sample=-1) = 0;
```

Comment:
Suggest avoiding the mix of virtual function and default parameters; it's one of those funky corners of C++: http://stackoverflow.com/questions/3533589/can-virtual-functions-have-default-parameters

An alternative is to provide a separate base overload, taking fewer parameters, which just calls the virtual function with the desired defaults. Then the derived class inherits that without having to repeat the defaults (or possibly change them)